sdk/core/Azure.Core/src/Internal/ArrayBackedPropertyBag.cs

```diff
@@ -11,33 +13,56 @@ namespace Azure.Core
     /// If the item count is less than 2, there are no allocations. Any additional items are stored in an array which will grow as needed.
     /// MUST be passed by ref only.
     /// </summary>
-    internal struct ArrayBackedPropertyBag
+    internal struct ArrayBackedPropertyBag<TKey, TValue> where TKey : struct, IEquatable<TKey>
     {
-        private KVP _first;
-        private KVP _second;
-        private KVP[] _rest;
+        private Kvp _first;
+        private Kvp _second;
+        private Kvp[] _rest;
         private int _count;
 
-        private readonly struct KVP
+        private readonly struct Kvp
         {
-            public readonly ulong Key;
-            public readonly object Value;
+            public readonly TKey Key;
+            public readonly TValue Value;
 
-            public KVP(ulong key, object value)
+            public Kvp(TKey key, TValue value)
             {
                 Key = key;
                 Value = value;
             }
+
+            public void Deconstruct(out TKey key, out TValue value)
+            {
+                key = Key;
+                value = Value;
+            }
         }
 
+        public int Count => _count;
+
         public bool IsEmpty => _count == 0;
 
-        public bool TryGetValue(ulong key, out object? value)
+        public void GetAt(int index, out TKey key, out TValue value)
+        {
+            if (index >= _count)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
```

Comment:
Should we avoid throwing directly to allow the method to be inlined?

Reply:
yeah, and we don't really need this exception, cause `_rest` would throw the same exception